async_fuse/src/memfs/cache/mod.rs

```diff
@@ -0,0 +1,587 @@
+use crate::fuse::fuse_reply::{AsIoVec, CouldBeAsIoVecList};
+use aligned_utils::bytes::AlignedBytes;
+use common::error::DatenLordResult;
+use libc;
+use lockfree_cuckoohash::{pin, LockFreeCuckooHash as HashMap};
+use log::debug;
+use nix::sys::uio::IoVec;
+use parking_lot::{Mutex, RwLock};
+use priority_queue::PriorityQueue;
+use std::ffi::OsStr;
+use std::fmt::{Debug, Error, Formatter};
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::os::unix::ffi::OsStrExt;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+/// Page Size
+const PAGE_SIZE: usize = 4096;
+/// The size of a memory block in KB, default is 64kB
+const MEMORY_BLOCK_SIZE_IN_KB: usize = 64;
+/// The size of a memory block in byte, default is 64kB
+const MEMORY_BLOCK_SIZE_IN_BYTE: usize = MEMORY_BLOCK_SIZE_IN_KB * 1024;
+/// The number of memory block in a Memory Bucket
+const MEMORY_BUCKET_VEC_SIZE: usize = 16;
+/// The size of a memory bucket in byte default is, 64 * 16 kB.
+#[allow(dead_code)]
+const MEMORY_BUCKET_SIZE_IN_BYTE: usize = MEMORY_BUCKET_VEC_SIZE * MEMORY_BLOCK_SIZE_IN_KB * 1024;
+/// The default capacity in bytes, 10GB
+const GLOBAL_CACHE_DEFAULT_CAPACITY: usize = 10 * 1024 * 1024 * 1024;
+
+/// A mapping from file name to per-file memory block mapping
+pub struct GlobalCache {
+    /// Map from file identifier to cache content map
+    inner: HashMap<Vec<u8>, HashMap<usize, MemBlockBucket>>,
+    /// Priority queue to track bucket usage
+    queue: Mutex<PriorityQueue<MemBlockBucket, i64>>,
```

Comment:
Do we need to use async mutex here?

Reply:
I'll modify the lock to async lock in latter commit. This commit just prooves the cache concept.